doc/coding_guidelines.md

```diff
@@ -0,0 +1,53 @@
+ - Easy transition to using contracts when they come to C++.
+ - Enable testing for pre-condition violations by conditionally throwing.
+
+
+## Heap allocations
+
+*snitch* code must not directly or indirectly allocate heap (or "free store") memory while running tests. This means that a number of common C++ STL classes cannot be used (at least not with the default allocator):
+ - `std::string`: use `std::string_view` (for constant strings) or `snitch::small_string` (for variable strings) instead.
+ - `std::vector`, `std::map`, `std::set`, and their variants: use `std::array` (for fixed size arrays) or `snitch::small_vector` (for variable size arrays) instead.
+ - `std::function`: use `snitch::small_function` instead.
+ - `std::unique_ptr`, `std::shared_ptr`: use values on the stack, and raw pointers for non-owning references.
+
+Unfortunately, the standard does not generally specify if a function or class allocates heap memory or not. We can make reasonable guesses for simple cases; in particular the following are fine to use:
+ - `std::string_view`,
+ - `std::array`,
+ - `std::span`,
+ - `std::variant` with `std::get_if` and `std::visit`,
+ - `std::optional`,
+ - `std::tuple`,
+ - Functions in `<algorithm>` ([except `std::stable_sort`, `std::stable_partition`, and `std::inplace_merge`](https://stackoverflow.com/a/46714875/1565581)).
+ - Concepts and type traits.
+
+Any type or function not listed above *should* be assumed to use heap memory unless demonstrated otherwise. One way to monitor this is on Linux is to use [valgrind/massif](https://valgrind.org/docs/manual/ms-manual.html).
+
+Note: If necessary, it is acceptable to allocate/de-allocate heap memory when a test is not running, i.e., either at the start or end of the program, or (if single-threaded) in between test cases. For example, as of writing this, with GCC 10 on Linux and the default reporter, *snitch* performs heap allocations on two occasions:
+ - several allocations at program startup, generated by [`libstdc++` initialisation](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=68606).
+ - one at the first console output, generated by `glibc` for its internal I/O buffer.
+
+
+## Heavy headers and compilation time
+
+One of the advantages of *snitch* over competing testing framework is fast compilation of tests. To preserve this advantage, ["heavy" STL headers](https://artificial-mind.net/projects/compile-health/) should not be included in *snitch* headers unless absolutely necessary. However, they can be included in the *snitch* implementation `*.cpp` files.
+
+Therefore:
+ - Place as much code as possible in the `*.cpp` files rather than in headers.
+ - When not possible (templates, constexpr, etc.), consider if you can use a short and clear hand-written alternative instead. For example, `std::max(a, b)` requires `<algorithm>`, but can also be written as `a > b ? a : b`. Some of the simplest algorithms in `<algorithm>`, like `std::copy`, can also be written with an explicit loop.
```

Comment:
```diff
-and clear hand-written alternative instead
+and clear handwritten alternative instead
```
[ngrams](https://books.google.com/ngrams/graph?content=hand-written%2Chandwritten&year_start=1800&year_end=2019&corpus=en-2019&smoothing=3)

Reply:
The more you know :) Thanks